### ISSUE TITLE
e2fsck - no progress shown

### ISSUE THREAD

**Issue by jury68000** [freemint/freemint#73]
When the ext2 partition is not unmounted correctly e2fsck being fired at boot stopped showing any progress. It looks like the booting progress hangs up, but it seems to scan the partition properly as when I wait few minutes the boot of MiNT starts to continue and after proper close of the system the next time running it doesnt fire up e2fsck.
![e2fsck](https://user-images.githubusercontent.com/3062260/43089609-656ddff2-8ea5-11e8-8c5c-98f6df0f7686.jpg)

**Comment by wongck68** [comment]
Normally for e2fsck during bootup, I add the -C 0 parameter so that it display a running progress bar.
You may have also accidentally redirected output from console to the captured log file in c:\mint.

**Comment by jury68000** [comment]
Thanks for the suggestion, but I'm not powered enough to try change those bash scripts, so for sure I did no change. But checked the script and there seems to be -C o parameter. 
The block triggering ext2 file system check looks like:

	for drv in $ext2drives; do
		echo; echo "Checking filesystem on drive $drv."
		fsck.ext2 -p -C 0 $drv:
		if [ $? -ge 2 ]; then
			echo
			echo "Warning: e2fsck FAILED ($?)"
			echo "         Starting single user shell. Fix the"
			echo "         filesystem on drive $drv by running fsck.ext2"
			echo "         and REBOOT!"
			echo
			sh -si < /dev/console > /dev/console 2>&1
		fi

**Comment by mikrosk** [comment]
@jury68000 can you post output of `fsck.ext2 -v` ? I've just tried to install old EasyMiNT and replace it with newest kernel and EmuTOS and it seems to work fine. So it's quite possible you're using a newer version of the tool and that causes some troubles.

**Comment by jury68000** [comment]
OK, added -v to the command from that script. Now its fsck.ext2 -v -p -C 0 $drv:
Then pressed the reset button and nothing changed.

**Comment by mikrosk** [comment]
No no, I meant just run it from the command  line. :) `fsck.ext2 -v`, nothing else.

**Comment by jury68000** [comment]
:)
Well, thats what I first done, but didn't see anything useful, so I thought that I understood it wrong and just changed the command in the script firing the fsck.
OK, so here is the output of your command:
```

Usage: fsck.ext2 [-panyrcdfvtDFV] [-b superblock] [-B blocksize]
		[-I inode_buffer_blocks] [-P process_inode_size]
		[-l|-L bad_blocks_file] [-C fd] [-j external_journal]
		[-E extended-options] device

Emergency help:
 -p                   Automatic repair (no questions)
 -n                   Make no changes to the filesystem
 -y                   Assume "yes" to all questions
 -c                   Check for bad blocks and add them to the badblock list
 -f                   Force checking even if filesystem is marked clean
 -v                   Be verbose
 -b superblock        Use alternative superblock
 -B blocksize         Force blocksize when looking for superblock
 -j external_journal  Set location of the external journal
 -l bad_blocks_file   Add to badblocks list
 -L bad_blocks_file   Set badblocks list
```

**Comment by mfro0** [comment]
The linked screenshot above shows an issue that has - as far as I can tell - nothing to do with `fsck`. It's an issue nevertheless.

What you show is the attempt to load the slip driver. That tries to open a named pipe with the `O_GLOBAL` flag which is obsolete. Because of that, the kernel answers with `EINVAL` and the driver exits.

That might (or might not) relate to your problem. It looks as if the kernel is mounting a file system at a place where it's not supposed to do that (maybe because of some error logging attempt), but I would recommend to disable/move away the slip driver (that isn't functional anymore anyway, but that's another issue) and try again.

**Comment by mikrosk** [comment]
I don't think the fsck disappearing is related to `slip.xif`: ![snap001](https://user-images.githubusercontent.com/3062260/43049025-7a2c1abc-8df1-11e8-9789-8ddc1dc03951.png). Personally I remember seeing this error also in EmuTOS+Aranym but for a world I can't replicate this bug again.

As for `O_GLOBAL`, it has been removed by this commit: https://github.com/freemint/freemint/commit/25294f68ed58fe2a4a73529a6017839bcbeb1d20, some explanation if also found here: https://mikro.naprvyraz.sk/mint/201312/msg00144.html. And yes, I didn't include it because of that error.

**Comment by mfro0** [comment]
You do not have the "mounting unchecked FS" warning in your screenshot that can be seen in the OP's one.

So there is something different...

relating slip: maybe the sources should be removed if they are not usable anymore?

**Comment by mfro0** [comment]
another idea: are we just fooled by the xconout2 output catcher?

**Comment by th-otto** [comment]
>maybe the sources should be removed if they are not usable anymore?

I think it's better to just don't build it, or distribute the binary. Although maybe rarely in use, there should be some way to make it work again.

**Comment by mikrosk** [comment]
I've taken another look at the screenshot. The fact I don't have the "mounting unchecked fs" (nor the "O_GLOBAL is obsolete, please update your driver (u:\pipe\sld)" for that matter) isn't an issue here (it's written to Aranym console) but something else is.

Don't you find intriguing that the latest module loaded is FAT/VFAT/FAT32? Where's the ram disk? Random number generator? xconout2? And most importantly, where the hell is Ext2 file system driver?!

Something switches off stdout output after the "AKP 0, ISO 0" message (stderr seems to be working just fine as we can see).

**Comment by jury68000** [comment]
Unsorted list of drivers:
![drivers](https://user-images.githubusercontent.com/32705667/43086457-a2a3a33c-8e9d-11e8-9dfc-033e10a3d3e3.jpg)

**Comment by mfro0** [comment]
Thank you.

No idea. But right now, I've seen the exact same behaviour in ARAnyM...

**Comment by zabruder** [comment]
---------------------------------------------------
> Thank you.
> 
> No idea. But right now, I've seen the exact same behaviour in ARAnyM...

I've never seen such thing.

Maybe silly, but did you check the boot-log?

**Comment by mfro0** [comment]
A real mess. For the same startup,

this is what ARAnyM has to show:
![grafik](https://user-images.githubusercontent.com/18102882/43088289-18035628-8ea2-11e8-97a2-6b7c4584ab8b.png)

This is what the ARAnyM console (and the boot.log, almost identical, just some different display of the ESC codes) has:

```
You have turned off the memory protection.
This is not recommended and may not be
supported in the future.

Calibrating delay loop ... 283.44 BogoMIPS

 High quality random number generator
 courtesy of Theodore Ts'o.
 Copyright ▒ 1994-1998 Theodore Ts'o
 All Rights Reserved.
 See the file COPYRAND for details.

Initializing built-in domain ops ... done!
Loading external modules ...

 MiNT-Net TCP/IP 1.6 PL 1,
 1993-1996 by Kay Roemer.
 1997-1999 by Torsten Lang.
 2000-2010 by Frank Naumann.

Loading interfaces:
Native Features Ethernet driver v0.7 ( )
IP  masquerading by Mario Becroft, 1999.
FTP masquerading support by Torsten Lang, 1999.

 xconout2 output catcher version 0.3
▒ Jan  1 1995 by TeSche <itschere@techfak.uni-bielefeld.de>
▒ 2000-2010 by Frank Naumann.

 NF_STDERR native feature output device version 0.1
▒ Jan  25 2005 by STanda Opichal <opichals@seznam.cz>
▒                 build date Apr 11 2013.

 Ext2 filesystem driver version 0.63
▒ 1998, 1999 by Axel Kaiser.
▒ 2000-2010 by Frank Naumann.

 Network file system driver version 0.57
▒ 1993, 1994 by Ulrich K▒hn.
▒ 2000-2010 by Frank Naumann.
See the file COPYING for copying and using conditions.

 HostFS Filesystem driver version 0.041
▒ Apr 11 2013 by ARAnyM Team
 WARNING: This is a test version - BETA!

Mounts: u:\h


```

**Comment by mfro0** [comment]
> @mfro0 so what happens if you disable boot.log ?

I *think* the problem goes away if I do so.

I'm on a different machine right now. This one hangs on XaAES startup if I *enable* `boot.log` 🤔

I will check on the same machine/installation that showed the strange behaviour tomorrow.

**Comment by mfro0** [comment]
Confirmed. Disabling boot.log at startup shows all boot messages (including the file system check progress bar) for me.

![grafik](https://user-images.githubusercontent.com/18102882/43124324-2162fc2a-8f27-11e8-939c-26a63d418919.png)

**Comment by mikrosk** [comment]
Now we need confirmation from @jury68000 whether this is the case for him. If so, I guess the solution is to fix the stdout to write to both channels, the log and console.

**Comment by zabruder** [comment]
---------------------------------------------------
> Now we need confirmation from @jury68000 whether this is the case for him. If so, I guess the solution is to fix the stdout to write to both channels, the log and console.

The aranym-MiNT-version already has 3 options for the boot-log: 
none/file/both (at least in my branch). Maybe that would work for the 
other targets also. If the file-only-method is selected, there could 
be an information like "logging to ...".

**Comment by mikrosk** [comment]
I guess the big part of the problem is that even if the 'file' option is selected, it still spits stderr to screen, eluding the user that something specific (like fsck) is broken and the rest is fine.

**Comment by zabruder** [comment]
Well, stderr shouldn't be too hard to add:

      rootproc->p_fd->ofiles[2] = fb;

?

**Comment by mfro0** [comment]
I don't think we should try to fix this with a hack but rather with a clean solution.

I would assume its already there (since some drivers appear to cope well with the redirect), it's just a matter of finding out how.

**Comment by th-otto** [comment]
@zabruder your hack just unconditionally redirects stderr. There is a reason why there are options for it. The current solution works just fine, i usually run aranym this way (logging to both). I think its not even an error for the messages of fsck to not appear: if you requested to write it only to boot.log, then it works as expected. The only problem might be when fsck runs interactively, and you don't see its questions (or realize that it might be waiting for input), because the output didn't appear on screen.

I think its more the other way around: the bug is that there is output from some drivers, but not from others. Maybe caused by the missing define mfro mentioned, in those drivers.

**Comment by mfro0** [comment]
O.k. It appears we first need to agree on what boot messages redirection is supposed to do?

1. write everything to a file and do not show anything onscreen for a "clean" boot experience
2. always show everything on screen but aditionally write it to a file (or device) to be able to diagnose later

I can live with both (while having a tendency to 2.), but at least it should be consistent. And if we go with 1., I'd suppose startup script messages should appear in all cases.

**Comment by zabruder** [comment]
@th-otto : Like it does for stdout now. I understood @mikrosk this way. Still I don't see why it is a a hack. Please explain. What "options" do you mean?

If the Cconws-output is the problem, then those should appear on the console. I've never seen that.

@mfro0: I think your 2. is what I meant with "both".

Sure, interactive input may still be an issue.

**Comment by zabruder** [comment]
---------------------------------------------------
> @zabruder i'm speaking about the WRITE_BOOT=x option from mint.ini. It already allows to choose where boot message should appear (only screen, only boot.log, or both)

And how is that related to stdout/stderr?

Would you suggest to add another option for the boot-menu to also 
redirect stderr or not? I don't think that is a good idea.

**Comment by jury68000** [comment]
> Now we need confirmation from @jury68000 whether this is the case for him. If so, I guess the solution is to fix the stdout to write to both channels, the log and console.

Yes, I confirm that after disabling boot.log the progress for fs scan shows properly.

**Comment by zabruder** [comment]
I guess the issue can be closed now, nothing special to e2fsck.

**Comment by th-otto** [comment]
Not special in the sense that there is something wrong with MiNT, but e2fsck actually *is* a bit special. First off, if something is wrong with the filesystem, so it cannot be mounted, we should make sure that at least that error message is written to the screen (could maybe be done without modifications to e2fsck, by checking the exit status in the script that runs it). 
The other problem is that we should make sure that something appears on screen when fsck tries to get answers from the user. If the prompt is only written to the boot.log, then it would look like the machine hangs. In that sense fsck is special, because it is maybe the only tool that might wait for user input.

(reopened again, if you still think this is not an issue, just close it again ;)

**Comment by mfro0** [comment]
The whole thing is not about `fsck` at all. IMHO, `fsck`is a userland program and we do not need to care about it in the kernel department.

For me, it's about inconsistent behaviour of kernel drivers during load. And it starts getting somehow personal for me, as I wasn't able to track down the reason for this yet ;)

**Comment by mikrosk** [comment]
I'm getting a bit lost in this debate. So we have three options regarding the boot log:
- do not log anything (easy)
- log to file only (kind of broken because some output is visible, some isn't)
- log to file and screen (broken for step-by-step as per @mfro0's report)

Apart from that we have a problem with user programs like `fsck` in case of 'file only' because they may require user intervention. 

Personally I'd be fine with logging only to file for 100% (i.e. not showing anything since as early as `mint.ini` has been parsed), fix the step-by-step problem and most importantly, reset/cancel/restore the redirection just before executing `INIT=` or `GEM=` part. XaAES has its own log and all user scripts can handle loggging on their own. So it's basically only a question of `/sbin/init` printing a message or two which wouldn't be logged.

What do you think?

**Comment by zabruder** [comment]
---------------------------------------------------
> Personally I'd be fine with logging only to file for 100% (i.e. not showing anything since as early as `mint.ini` has been parsed), fix the step-by-step problem and most importantly, reset/cancel/restore the redirection just before executing `INIT=` or `GEM=` part. XaAES has its own log and all user scripts can handle loggging on their own. So it's basically only a question of `/sbin/init` printing a message or two which wouldn't be logged.
> 
> What do you think?

The redirection of course gets restored before INIT=/GEM=. Do you 
notice something different?

XaAES re-uses the boot-log because it uses the cnf-parser from MiNT (I
think), but that's another topic.

**Comment by mikrosk** [comment]
> The redirection of course gets restored before INIT=/GEM=. Do you notice something different?

Oops, my mistake. I didn't realise `fsck` is executed from `mint.cnf`, not as a user script. In that case, hmm... maybe we should eliminate the whole "file only" approach as whole.

**Comment by zabruder** [comment]
---------------------------------------------------
> > The redirection of course gets restored before INIT=/GEM=. Do you notice something different?
> 
> Oops, my mistake. I didn't realise `fsck` is executed from `mint.cnf`, not as a user script. In that case, hmm... maybe we should eliminate the whole "file only" approach as whole.

Er - why? Maybe it's better to run fsck from init. BTW: fsck never 
reads user-input if called correctly.

But do as you like.

**Comment by mikrosk** [comment]
You've misunderstood - I am merely suggesting to remove one of the options (to log into file *only*, i.e. keeping file+screen as it used to be).

**Comment by zabruder** [comment]
---------------------------------------------------
> You've misunderstood - I am merely suggesting to remove one of the options (to log into file *only*, i.e. keeping file+screen as it used to be).

That only exists for aranym, and would require additional research for
real hardware (e.g. call TRAP_Cconws additionally?). And you could not
get rid of the sometimes annoying scrolling.

**Comment by wongck68** [comment]
LOL... as already mentioned back on 9 Feb, 2nd posting of this thread..... check the Boot.log file.

I also have this issue when I ( the user ) who specifically asked MInt to log all boots notifications to the file. It is NOT a Mint issue nor a FSCK issue.... but an issue between the keyboard and the chair. if the User forgot he specifically asked for logging in to the boot.log, then there is nothing we can do.

I also have set keypress to continue with BOOT LOG and had been waiting for along time for Mint to boot up.... because I forgot and nothing on the screen, no HDD lights blinking.... Panic mode because poor Falcon/CT63 may be dead.... but it is not.... just waiting for USER ( me) to press a key to continue.

What I thought back then when I realised my dumb forgetfulness is

 " why can't MINT just tell me that it is now logging into a file and please press a key to continue."

So IMHO just do these modifications will do:
1/ information sentence stating.... "logging to boot.log is on/enabled" as a reminder to forgetful user.
2/ instead of printing "press a key to continue" into the boot.log, print it on the screen (stderr related in Cconws??)

**Comment by th-otto** [comment]
>Panic mode because poor Falcon/CT63 may be dead.

Yes, that's what i also think is the somewhat bad consequence of it. I think this happened to me when configuring aranym without nfjpeg support, then the jpeg driver will also wait for keypress. Grmpf.

>information sentence stating.... "logging to boot.log is on/enabled" as a reminder to forgetful user.

Hm, i think that would defeat the purpose of "log everything to boot.log and don't bother me with anything on screen".

>Instead of printing "press a key to continue" into the boot.log, print it on the screen

No idea how you want to achieve that. Mint cannot know which message (if any at all) is printed before the program waits for a key. And checking stdin for that purpose might not work either, if the program uses Bconin() instead of some gemdos function.

**Comment by zabruder** [comment]
> No idea how you want to achieve that. Mint cannot know which message (if any at all) is printed before the program waits for a key. And checking stdin for that purpose might not work either, if the program uses Bconin() instead of some gemdos function.

Maybe in Cconin it could print something on screen when write_boot != 
0.

**Comment by wongck68** [comment]
No... the "press key to continue" is something Mint put on the screen after each loading of XFS, XDD etc.
So MInt knows when it happens. 
I am not stating the output from XDD, XFS etc as for sure it is inside the XDD code that must output ALERT(..) or INFO(..) or a new #define for 100% output to screen.

**Comment by wongck68** [comment]
> Hm, i think that would defeat the purpose of "log everything to boot.log and don't bother me with anything on screen".

I am just asking for a one-liner before the logging starts. In any case, there was still some notices that appears even when boot-log is enabled... or am i mistaken ( as normally )

**Comment by wongck68** [comment]
The solution is apparent for this boot-logging blank screen with "press key to continue" issue. The boot logging and "press key to continue" just need to be mutually exclusive.

**Comment by mikrosk** [comment]
Hilarious. I was doing some last changes for #356 and bumped exactly into this issue: no progress bar shown.

Fortunately, it was about missing `-C 0` but for a while I was really worried.

Anyway, this task is still very valid, that redirection into file and/or interactive mode has to be solved.
